backend/models/user.js

```diff
@@ -74,12 +91,57 @@ const findOne = async function (query, projection) {
 	return await DB.findOne("admin", COLL_NAME, query, projection);
 };
 
+const handleAuthenticateFail = async function (username) {
+	const currentTime = Date.now();
+
+	try {
+		const user = await User.findByUserName(username);
+		const elapsedTime = user.customData.lastFailedLoginAt ?
+			currentTime - user.customData.lastFailedLoginAt : undefined;
+
+		const failedLoginCount = user.customData.failedLoginCount && elapsedTime && elapsedTime < config.lockoutDuration ?
+			user.customData.failedLoginCount + 1 : 1;
+
+		const loginLocked = failedLoginCount >= config.maxUnsuccessfulLoginAttempts ? true : undefined;
+
+		await User.update(username, {
+			"customData.lastFailedLoginAt": currentTime,
+			"customData.failedLoginCount": failedLoginCount,
+			"customData.loginLocked": loginLocked
+		});
+
+		return Math.max(config.maxUnsuccessfulLoginAttempts - failedLoginCount, 0);
+	} catch(err) {
+		// suppress update failure
+	}
+};
+
 const COLL_NAME = "system.users";
 
 const User = {};
 
 User.update = async function (username, data) {
-	return DB.update("admin", COLL_NAME, {user: username}, {$set: data});
+	const toUpdate = {};
```

Comment:
we should really be ditching this function instead of expanding on it, 

In this case it doesn't even do type checking, so it's quite useless